src/java/org/apache/cassandra/db/marshal/VectorType.java

```diff
@@ -0,0 +1,613 @@
+import org.apache.cassandra.db.TypeSizes;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.serializers.MarshalException;
+import org.apache.cassandra.serializers.TypeSerializer;
+import org.apache.cassandra.transport.ProtocolVersion;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.JsonUtils;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+
+public final class VectorType<T> extends AbstractType<List<T>>
+{
+    private static class Key
+    {
+        private final AbstractType<?> type;
+        private final int dimension;
+
+        private Key(AbstractType<?> type, int dimension)
+        {
+            this.type = type;
+            this.dimension = dimension;
+        }
+
+        private VectorType<?> create()
+        {
+            return new VectorType<>(type, dimension);
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Key key = (Key) o;
+            return dimension == key.dimension && Objects.equals(type, key.type);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(type, dimension);
+        }
+    }
+    @SuppressWarnings("rawtypes")
+    private static final ConcurrentHashMap<Key, VectorType> instances = new ConcurrentHashMap<>();
+
+    public final AbstractType<T> elementType;
+    public final int dimension;
+    private final TypeSerializer<T> elementSerializer;
+    private final int valueLengthIfFixed;
+    private final VectorSerializer serializer;
+
+    private VectorType(AbstractType<T> elementType, int dimension)
+    {
+        super(ComparisonType.CUSTOM);
+        if (dimension <= 0)
+            throw new InvalidRequestException(String.format("vectors may only have positive dimentions; given %d", dimension));
+        this.elementType = elementType;
+        this.dimension = dimension;
+        this.elementSerializer = elementType.getSerializer();
+        this.valueLengthIfFixed = elementType.isValueLengthFixed() ?
+                                  elementType.valueLengthIfFixed() * dimension :
+                                  super.valueLengthIfFixed();
+        this.serializer = elementType.isValueLengthFixed() ?
+                          new FixedLengthSerializer() :
+                          new VariableLengthSerializer();
+    }
+
+    public static <T> VectorType<T> getInstance(AbstractType<T> elements, int dimension)
+    {
+        Key key = new Key(elements, dimension);
+        return instances.computeIfAbsent(key, Key::create);
+    }
+
+    public static VectorType<?> getInstance(TypeParser parser)
+    {
+        TypeParser.Vector v = parser.getVectorParameters();
+        return new VectorType<>(v.type, v.dimension);
```

Comment:
This could use the cached instances:
```suggestion
        return getInstance(v.type, v.dimension);
```

Reply:
thanks, keep forgetting that the constructor in `AbstractType` uses reflection =(